File: src/complydoc/extraction/routing.py
"""Which extraction path each page needs: its text layer, OCR, or a vision model.

    plan = plan_routes(document, config.readiness.routing)
    plan.counts  # {"text": 8, "ocr": 2, "vision": 1}

Sending every page to a vision model reads anything and costs the most; reading
the text layer costs the least and returns nothing for a scan. Most folders are a
mix, and the mix is decided page by page:

| Route | When |
| --- | --- |
| `text` | A usable text layer, and nothing on the page that plain text loses |
| `ocr` | No usable text layer, and a scan OCR can read |
| `vision` | Plain text would lose the page |

A page takes the vision route when it carries a table with merged or stacked
header cells, when it is mostly picture with a caption for a text layer, when it
is a scan too coarse for OCR, or when OCR read it poorly.

Each page carries the reason for its route, so a plan can be argued with. The
thresholds are the ones in `readiness.yaml`, under `routing`.

Routes are decided while the document is open, because per-page tables, scan
resolution and OCR confidence are not carried in the report.
"""

from __future__ import annotations

from dataclasses import dataclass, field

from complydoc.config.schema import RoutingConfig
from complydoc.ingest.base import Document, Page
from complydoc.utils.geometry import coverage_fraction

__all__ = ["ROUTES", "DocumentRouting", "PageRoute", "plan_routes"]

ROUTES = ("text", "ocr", "vision")
"""Cheapest first."""


@dataclass(frozen=True, slots=True)
class PageRoute:
    number: int
    route: str
    """One of `ROUTES`."""
    reason: str
    """Why this page takes this route."""
    characters: int
    """Characters of text on the page, for apportioning the text-path cost."""


@dataclass(frozen=True, slots=True)
class DocumentRouting:
    pages: list[PageRoute] = field(default_factory=list)

    @property
    def counts(self) -> dict[str, int]:
        return {route: sum(1 for p in self.pages if p.route == route) for route in ROUTES}

    @property
    def route(self) -> str | None:
        """The most expensive route any page needs, which the document as a whole needs."""
        for route in reversed(ROUTES):
            if any(page.route == route for page in self.pages):
                return route
        return None


def _image_share(page: Page) -> float:
    if page.area_pt <= 0:
        return 0.0
    return coverage_fraction([b.bbox for b in page.image_blocks], page.width_pt, page.height_pt)


def _text_share(page: Page) -> float:
    if page.area_pt <= 0 or not page.text_blocks:
        return 0.0
    return coverage_fraction([b.bbox for b in page.text_blocks], page.width_pt, page.height_pt)


def _awkward_table(page: Page) -> bool:
    """A table whose shape is the information: merged cells, or stacked headers."""
    return any(table.merged_cells > 0 or table.header_depth > 1 for table in page.tables)


def _page_route(page: Page, settings: RoutingConfig) -> tuple[str, str]:
    readable = page.text_source in ("native", "loader") and len(page.text.strip()) >= (
        settings.min_characters
    )
    image_share = _image_share(page)
    mostly_picture = image_share * 100 >= settings.picture_share_pct

    if readable:
        if settings.vision_for_complex_tables and _awkward_table(page):
            return "vision", "a table with merged or stacked header cells, which plain text loses"
        text_share = _text_share(page)
        if mostly_picture and text_share * 100 < settings.min_text_coverage_pct:
            return (
                "vision",
                f"mostly picture ({image_share * 100:.0f}% of the page) with "
                f"{text_share * 100:.0f}% text, so the text layer is a caption",
            )
        return "text", f"a text layer covering {text_share * 100:.0f}% of the page"

    if page.ocr_confidence is not None and page.ocr_confidence * 100 < settings.min_ocr_confidence:
        return (
            "vision",
            f"OCR read this page with {page.ocr_confidence * 100:.0f}% confidence, "
            f"below {settings.min_ocr_confidence:g}%",
        )

    dpi = page.estimated_dpi() if mostly_picture else None
    if dpi is not None and dpi < settings.min_ocr_dpi:
        return "vision", f"a scan at {dpi:.0f} dpi, below the {settings.min_ocr_dpi:g} OCR needs"
    if mostly_picture:
        where = f" at {dpi:.0f} dpi" if dpi is not None else ""
        return "ocr", f"a scan{where} with no text layer"
    return "ocr", "no usable text layer, so the page has to be recognised"


def plan_routes(document: Document, settings: RoutingConfig | None = None) -> DocumentRouting:
    """The route each page of `document` needs, with the reason for each."""
    settings = settings or RoutingConfig()
    return DocumentRouting(
        pages=[
            PageRoute(
                number=page.number,
                route=route,
                reason=reason,
                characters=len(page.text),
            )
            for page in document.pages
            for route, reason in [_page_route(page, settings)]
        ]
    )

File: landing/src/data/routing.ts
// From `complydoc routing ./documents` on viewer/sample/documents (complydoc-routing.json),
// priced for Claude Sonnet 5 at medium resolution. Rerun it rather than editing this file.

export type Route = "text" | "ocr" | "vision";

export interface RoutedPage {
  page: number;
  route: Route;
  reason: string;
  characters: number;
}

export const ROUTING_MODEL = "Claude Sonnet 5";

export const ROUTING_COST_USD = {"routed": 0.047748, "text_layer": 0.035932, "text_ocr": 0.035932, "vision": 0.1056};

export const ROUTED_DOCUMENTS: { document: string; pages: RoutedPage[] }[] = [
  {
    "document": "annual-report-2025.pdf",
    "pages": [
      {
        "page": 1,
        "route": "text",
        "reason": "a text layer covering 48% of the page",
        "characters": 3838
      },
      {
        "page": 2,
        "route": "text",
        "reason": "a text layer covering 17% of the page",
        "characters": 1336
      },
      {
        "page": 3,
        "route": "vision",
        "reason": "a table with merged or stacked header cells, which plain text loses",
        "characters": 2412
      },
      {
        "page": 4,
        "route": "text",
        "reason": "a text layer covering 54% of the page",
        "characters": 4393
      },
      {
        "page": 5,
        "route": "text",
        "reason": "a text layer covering 54% of the page",
        "characters": 4401
      },
      {
        "page": 6,
        "route": "text",
        "reason": "a text layer covering 53% of the page",
        "characters": 4354
      },
      {
        "page": 7,
        "route": "text",
        "reason": "a text layer covering 53% of the page",
        "characters": 4368
      },
      {
        "page": 8,
        "route": "text",
        "reason": "a text layer covering 53% of the page",
        "characters": 4339
      }
    ]
  },
  {
    "document": "employee-handbook.pdf",
    "pages": [
      {
        "page": 1,
        "route": "text",
        "reason": "a text layer covering 51% of the page",
        "characters": 5143
      },
      {
        "page": 2,
        "route": "text",
        "reason": "a text layer covering 56% of the page",
        "characters": 5694
      },
      {
        "page": 3,
        "route": "text",
        "reason": "a text layer covering 56% of the page",
        "characters": 5631
      },
      {
        "page": 4,
        "route": "text",
        "reason": "a text layer covering 57% of the page",
        "characters": 5711
      },
      {
        "page": 5,
        "route": "text",
        "reason": "a text layer covering 17% of the page",
        "characters": 1759
      },
      {
        "page": 6,
        "route": "text",
        "reason": "a text layer covering 7% of the page",
        "characters": 549
      }
    ]
  },
  {
    "document": "master-services-agreement.pdf",
    "pages": [
      {
        "page": 1,
        "route": "text",
        "reason": "a text layer covering 27% of the page",
        "characters": 2053
      },
      {
        "page": 2,
        "route": "text",
        "reason": "a text layer covering 50% of the page",
        "characters": 4148
      },
      {
        "page": 3,
        "route": "text",
        "reason": "a text layer covering 50% of the page",
        "characters": 4165
      },
      {
        "page": 4,
        "route": "text",
        "reason": "a text layer covering 52% of the page",
        "characters": 4230
      },
      {
        "page": 5,
        "route": "text",
        "reason": "a text layer covering 52% of the page",
        "characters": 4298
      },
      {
        "page": 6,
        "route": "text",
        "reason": "a text layer covering 53% of the page",
        "characters": 4405
      },
      {
        "page": 7,
        "route": "text",
        "reason": "a text layer covering 14% of the page",
        "characters": 1195
      },
      {
        "page": 8,
        "route": "text",
        "reason": "a text layer covering 6% of the page",
        "characters": 451
      }
    ]
  },
  {
    "document": "rechnungen-2026-de.pdf",
    "pages": [
      {
        "page": 1,
        "route": "text",
        "reason": "a text layer covering 9% of the page",
        "characters": 582
      },
      {
        "page": 2,
        "route": "text",
        "reason": "a text layer covering 9% of the page",
        "characters": 583
      },
      {
        "page": 3,
        "route": "text",
        "reason": "a text layer covering 9% of the page",
        "characters": 580
      }
    ]
  },
  {
    "document": "supplier-invoices-scanned.pdf",
    "pages": [
      {
        "page": 1,
        "route": "vision",
        "reason": "a scan at 150 dpi, below the 200 OCR needs",
        "characters": 0
      },
      {
        "page": 2,
        "route": "vision",
        "reason": "a scan at 150 dpi, below the 200 OCR needs",
        "characters": 0
      },
      {
        "page": 3,
        "route": "vision",
        "reason": "a scan at 150 dpi, below the 200 OCR needs",
        "characters": 0
      }
    ]
  },
  {
    "document": "vendor-due-diligence.pdf",
    "pages": [
      {
        "page": 1,
        "route": "text",
        "reason": "a text layer covering 24% of the page",
        "characters": 1967
      },
      {
        "page": 2,
        "route": "text",
        "reason": "a text layer covering 15% of the page",
        "characters": 1361
      },
      {
        "page": 3,
        "route": "text",
        "reason": "a text layer covering 13% of the page",
        "characters": 1201
      },
      {
        "page": 4,
        "route": "text",
        "reason": "a text layer covering 14% of the page",
        "characters": 1217
      },
      {
        "page": 5,
        "route": "text",
        "reason": "a text layer covering 4% of the page",
        "characters": 283
      }
    ]
  }
];
